test/zdtm/static/Makefile

```diff
@@ -54,6 +54,7 @@ TST_NOFILE	:=				\
 		shm-mp				\
 		ptrace_sig			\
 		pidfd_self			\
+		pidfd_dead			\
```

Comment:
It looks like this test fails in Fedora Rawhide:

```
======================= Run zdtm/static/pidfd_dead in h ========================
Start test
./pidfd_dead --pidfile=pidfd_dead.pid --outfile=pidfd_dead.out
Run criu dump
Run criu restore
=[log]=> dump/zdtm/static/pidfd_dead/168/1/restore.log
------------------------ grep Error ------------------------
b'(00.004951)    168: \t\t\tGoing to dup 1 into 2'
b'(00.004953)    168: \tReceive fd for 2'
b'(00.004961)    168: \t\tCreate fd for 3'
b'(00.005264)    168: \t\tCreate fd for 4'
b'(00.005486)    168: Error (criu/cr-restore.c:1261): 175 killed by signal 9: Killed'
b'(00.005504) Error (criu/cr-restore.c:2314): Restoring FAILED.'
------------------------ ERROR OVER ------------------------
############### Test zdtm/static/pidfd_dead FAIL at CRIU restore ###############
Test output: ================================
```
https://github.com/checkpoint-restore/criu/runs/31040419148

Reply:
I will try to see why this happens!
Is there any way that I can replicate this test locally?
I did try running the tests on a VM with Fedora Rawhide `6.12.0-0.rc1.20241005git27cc6fdf7201.22.fc42.x86_64`
where these tests end up passing!

Reply:
@bsach64 There is a race condition between `kill()` and `waitpid()` in `free_dead_pidfd()`.

You can replicate the error with the following change:
```diff
diff --git a/criu/pidfd.c b/criu/pidfd.c
index fdf5dec60..23062ae02 100644
--- a/criu/pidfd.c
+++ b/criu/pidfd.c
@@ -151,7 +151,7 @@ static int free_dead_pidfd(struct dead_pidfd *dead)
                dead->pid);
                goto err;
        }
-
+       sleep(1);
        if (waitpid(dead->pid, &status, 0) != dead->pid) {
                pr_perror("Could not wait on temporary process with pid: %d",
                dead->pid);

```

```
$ sudo python3 zdtm.py run -t zdtm/static/pidfd_dead
userns is supported
=== Run 1/1 ================ zdtm/static/pidfd_dead
====================== Run zdtm/static/pidfd_dead in uns =======================
Start test
./pidfd_dead --pidfile=pidfd_dead.pid --outfile=pidfd_dead.out
Run criu dump
Run criu restore
=[log]=> dump/zdtm/static/pidfd_dead/64/1/restore.log
------------------------ grep Error ------------------------
b'(00.016134)      1: No ipcns-sem-11.img image'
b'(00.020782)      1: net: Try to restore a link 10:1:lo'
b'(00.020820)      1: net: Restoring link lo type 1'
b'(00.024435)      1: net: \tRunning ip addr restore'
b'Error: ipv4: Address already assigned.'
b'Error: ipv6: address already assigned.'
b'(00.141528)      4: \t\tCreate fd for 3'
b'(00.141715)      1: `- render 1 iovs (0x7f8433c61000:8192...)'
b'(00.141790)      1: Restore via sigreturn'
b'(00.142590)      4: \t\tCreate fd for 4'
b'(00.143514)      4: Error (criu/cr-restore.c:1261): 18 killed by signal 9: Killed'
b'(00.143579)      4: Error (criu/pidfd.c:156): pidfd: Could not wait on temporary process with pid: 18: No child processes'
b'(00.143591)      4: Error (criu/pidfd.c:218): pidfd: Failed to delete dead_pidfd struct'
b"(00.143605)      4: Error (criu/pidfd.c:234): pidfd: Can't create pidfd 0x00000d NSpid: -1 flags: 0"
b'(00.143615)      4: Error (criu/files.c:1221): Unable to open fd=5 id=0xd'
b'(00.147752) uns: calling exit_usernsd (-1, 1)'
b'(00.147902) uns: daemon calls 0x479de0 (91, -1, 1)'
b'(00.147939) uns: `- daemon exits w/ 0'
b'(00.150476) uns: daemon stopped'
b'(00.150514) Error (criu/cr-restore.c:2314): Restoring FAILED.'
------------------------ ERROR OVER ------------------------
############### Test zdtm/static/pidfd_dead FAIL at CRIU restore ###############
Test output: ================================

 <<< ================================
##################################### FAIL #####################################
```

Reply:
Another solution might be to block `SIGCHLD`?
See `sysctl.c` line 271 https://github.com/checkpoint-restore/criu/blob/criu-dev/criu/sysctl.c#L271